routing/bicycle_directions.cpp

```diff
@@ -222,39 +213,172 @@ Index::FeaturesLoaderGuard & BicycleDirectionsEngine::GetLoader(MwmSet::MwmId co
+  adjacentEdges.m_outgoingTurns.candidates.reserve(outgoingEdges.size());
+  uniNodeId = UniNodeId(inEdge.GetFeatureId(), startSegId, endSegId, inEdge.IsForward());
+  m2::PointD const & ingoingPoint = inEdge.GetStartJunction().GetPoint();
+  m2::PointD const & junctionPoint = inEdge.GetEndJunction().GetPoint();
+
+  for (auto const & edge : outgoingEdges)
+  {
+    if (edge.IsFake())
+      continue;
+
+    auto const & outFeatureId = edge.GetFeatureId();
+    FeatureType ft;
+    if (!GetLoader(outFeatureId.m_mwmId).GetFeatureByIndex(outFeatureId.m_index, ft))
+      continue;
+
+    auto const highwayClass = ftypes::GetHighwayClass(ft);
+    ASSERT_NOT_EQUAL(highwayClass, ftypes::HighwayClass::Error, ());
+    ASSERT_NOT_EQUAL(highwayClass, ftypes::HighwayClass::Undefined, ());
+
+    double angle = 0;
+
+    if (inEdge.GetFeatureId().m_mwmId == edge.GetFeatureId().m_mwmId)
+    {
+      ASSERT_LESS(MercatorBounds::DistanceOnEarth(junctionPoint, edge.GetStartJunction().GetPoint()),
+                  turns::kFeaturesNearTurnMeters, ());
+      m2::PointD const & outgoingPoint = edge.GetEndJunction().GetPoint();
+      angle = my::RadToDeg(turns::PiMinusTwoVectorsAngle(junctionPoint, ingoingPoint, outgoingPoint));
+    }
+    else
+    {
+      // Note. In case of crossing mwm border
+      // (inEdge.GetFeatureId().m_mwmId != edge.GetFeatureId().m_mwmId)
+      // twins of inEdge.GetFeatureId() are considered as outgoing features.
+      // In this case that turn candidate angle is invalid and
+      // should not be used for turn generation.
+      adjacentEdges.m_outgoingTurns.isCandidatesAngleValid = false;
+    }
+    adjacentEdges.m_outgoingTurns.candidates.emplace_back(angle, uniNodeId, highwayClass);
+  }
+}
+
+Segment BicycleDirectionsEngine::GetSegment(FeatureID const & featureId, uint32_t segId, bool forward) const
+{
+  CHECK(m_numMwmIds, ());
+
+  if (!featureId.m_mwmId.IsAlive())
+    MYTHROW(RoutingException, ("Mwm:", featureId.m_mwmId, "is not alive."));
+
+  NumMwmId const numMwmId =
+    m_numMwmIds->GetId(featureId.m_mwmId.GetInfo()->GetLocalFile().GetCountryFile());
+  return Segment(numMwmId, featureId.m_index, segId, forward);
+}
+
+void BicycleDirectionsEngine::FillPathSegmentsAndAdjacentEdgesMap(
+    RoadGraphBase const & graph, vector<Junction> const & path,
+    IRoadGraph::TEdgeVector const & routeEdges, my::Cancellable const & cancellable)
+{
+  size_t const pathSize = path.size();
+  CHECK_GREATER(pathSize, 1, ());
+  CHECK_EQUAL(routeEdges.size(), pathSize - 1, ());
+  // Filling |m_adjacentEdges|.
+  m_adjacentEdges.insert(make_pair(UniNodeId(UniNodeId::Type::Mwm), AdjacentEdges(0)));
+  auto constexpr kInvalidSegId = std::numeric_limits<uint32_t>::max();
+  // |startSegId| is a value to keep start segment id of a new instance of LoadedPathSegment.
+  uint32_t startSegId = kInvalidSegId;
+  vector<Junction> prevJunctions;
+  vector<Segment> prevSegments;
+  for (size_t i = 1; i < pathSize; ++i)
```

Comment:
Could you please rewrite this loop to a couple of loops: inner and outer, like this:
```c++
size_t i = 1;
while (i < pathSize) {
   size_t j = i;
   while (j < pathSize && IsNotJoint(j))
      ++j;
   i = j;
}
```
I mean, you implicitly do this, by accumulating non-joint segments to prevSegments and non-joint junctions to prevJunctions. From my point of view this is hard to read, why not to do that explicitly?